<launch>

  <arg name="robot" default="minitaur"/>
  <arg name="calibration_file"
    default="$(find legged_vio)/config/zed_imu_camera_altered.yaml"/>

  <rosparam command="load" file="$(arg calibration_file)"/>

  <!-- Image Processor Nodelet  -->
  <group ns="$(arg robot)">
    
    <node pkg="nodelet" type="nodelet" name="image_processor"
      args="standalone legged_vio/ImageProcessorNodelet"
      output="screen">
      
      <rosparam command="load" file="$(arg calibration_file)"/>
      <param name="grid_row" value="4"/>
      <param name="grid_col" value="5"/>
      <param name="grid_min_feature_num" value="3"/>
      <param name="grid_max_feature_num" value="4"/>
      <param name="pyramid_levels" value="3"/>
      <param name="patch_size" value="15"/>
      <param name="fast_threshold" value="10"/>
      <param name="max_iteration" value="30"/>
      <param name="track_precision" value="0.01"/>
      <param name="ransac_threshold" value="3"/>
      <param name="stereo_threshold" value="5"/>
      
      <!-- START: eventually move these lines to legged_vio_zed.launch -->
      <!-- These values should be standard deviation -->
      <param name="noise/gyro" value="0.005"/>
      <param name="noise/acc" value="0.05"/>
      <param name="noise/gyro_bias" value="0.001"/>
      <param name="noise/acc_bias" value="0.01"/>
      <param name="noise/feature" value="0.035"/>

      <param name="initial_state/velocity/x" value="0.0"/>
      <param name="initial_state/velocity/y" value="0.0"/>
      <param name="initial_state/velocity/z" value="0.0"/>

      <!-- These values should be covariance -->
      <param name="initial_covariance/velocity" value="0.25"/>
      <param name="initial_covariance/gyro_bias" value="0.01"/>
      <param name="initial_covariance/acc_bias" value="0.01"/>
      <param name="initial_covariance/extrinsic_rotation_cov" value="3.0462e-4"/>
      <param name="initial_covariance/extrinsic_translation_cov" value="2.5e-5"/>
      <!-- END: eventually move these lines to legged_vio_zed.launch -->

      <!-- This nodelet subscribes to imu, cam0_image, and cam1_image topics -->
      <!-- Remap to the topics that are published by stereo camera: /imu0/, /cam0/image_raw, etc -->
      <remap from="~imu" to="/zed/zed_node/imu/data_raw"/>
      <remap from="~cam0_image" to="/zed/zed_node/left/image_rect_color"/>
      <remap from="~cam1_image" to="/zed/zed_node/right/image_rect_color"/>

    </node>
  </group>

</launch>
